src/pocketmine/Player.php

```diff
@@ -3322,131 +3301,192 @@ public function handleDataPacket(DataPacket $packet){
 
-							foreach($needed as $k => $n){
-								if($n->deepEquals($item, $n->getDamage() !== null, $n->getCompoundTag() !== null)){
-									$remove = min($n->getCount(), $item->getCount());
-									$n->setCount($n->getCount() - $remove);
-									$item->setCount($item->getCount() - $remove);
+						if($ev->isCancelled()){
+							$this->inventory->sendContents($this);
+							break;
+						}
 
-									if($n->getCount() === 0){
-										unset($needed[$k]);
+						$this->floatingInventory = $floatingInventory; //Set player crafting inv to the idea one created in this process
+						$this->floatingInventory->addItem(clone $recipe->getResult()); //Add the result to our picture of the crafting inventory
+					}else{
+						$this->server->getLogger()->debug("Unmatched desktop crafting recipe " . $recipe->getId() . " from player " . $this->getName());
+						$this->inventory->sendContents($this);
+						break;
+					}
+				}else{
+					if($recipe instanceof ShapedRecipe){
+						for($x = 0; $x < 3 and $canCraft; ++$x){
+							for($y = 0; $y < 3; ++$y){
+								$item = $packet->input[$y * 3 + $x];
+								$ingredient = $recipe->getIngredient($x, $y);
+								if($item->getCount() > 0 and $item->getId() > 0){
+									if($ingredient == null){
+										$canCraft = false;
+										break;
 									}
+									if($ingredient->getId() != 0 and !$ingredient->deepEquals($item, $ingredient->getDamage() !== null, $ingredient->getCompoundTag() !== null)){
+										$canCraft = false;
+										break;
+									}
+
+								}elseif($ingredient !== null and $item->getId() !== 0){
+									$canCraft = false;
+									break;
 								}
 							}
+						}
+					}elseif($recipe instanceof ShapelessRecipe){
+						$needed = $recipe->getIngredientList();
 
-							if($item->getCount() > 0){
-								$canCraft = false;
-								break;
+						for($x = 0; $x < 3 and $canCraft; ++$x){
+							for($y = 0; $y < 3; ++$y){
+								$item = clone $packet->input[$y * 3 + $x];
+
+								foreach($needed as $k => $n){
+									if($n->deepEquals($item, $n->getDamage() !== null, $n->getCompoundTag() !== null)){
+										$remove = min($n->getCount(), $item->getCount());
+										$n->setCount($n->getCount() - $remove);
+										$item->setCount($item->getCount() - $remove);
+
+										if($n->getCount() === 0){
+											unset($needed[$k]);
+										}
+									}
+								}
+
+								if($item->getCount() > 0){
+									$canCraft = false;
+									break;
+								}
 							}
 						}
-					}
-
-					if(count($needed) > 0){
+						if(count($needed) > 0){
+							$canCraft = false;
+						}
+					}else{
 						$canCraft = false;
 					}
-				}else{
-					$canCraft = false;
-				}
 
-				/** @var Item[] $ingredients */
-				$canCraft = true;//0.13.1大量物品本地配方出现问题,无法解决,使用极端(唯一)方法修复.
-				$ingredients = $packet->input;
-				$result = $packet->output[0];
+					//Nasty hack. TODO: Get rid
+					$canCraft = true;//0.13.1大量物品本地配方出现问题,无法解决,使用极端(唯一)方法修复.
```

Comment:
Another dirty (~~chinese~~) hack from people too lazy to fix things the right way.